machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,247 @@
+// +kubebuilder:metadata:labels=openshift.io/operator-managed=
+
+// MachineOSConfig describes the configuration for a build process managed by the MCO
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineOSConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of the machineosconfig
+	// +kubebuilder:validation:Required
+	Spec MachineOSConfigSpec `json:"spec"`
+
+	// status describes the status of the machineosconfig
+	// +optional
+	Status *MachineOSConfigStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSConfigList describes all configurations for image builds on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineOSConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// items contains a collection of MachineOSConfig resources.
+	Items []MachineOSConfig `json:"items"`
+}
+
+// MachineOSConfigSpec describes user-configurable options as well as information about a build process.
+type MachineOSConfigSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
+	// buildInputs is where user input options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// buildOutputs holds all information needed to handle booting the image after a build
+	// This currently contains a currentImagePullSecret field, which should be provided if the final pull secret used to pull the image to nodes from the registry
+	// is different than the one used for pushing the image to the registry during the build.
+	// +optional
+	BuildOutputs *BuildOutputs `json:"buildOutputs,omitempty"`
+}
+
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
+	// observedGeneration represents the generation of the MachineOSConfig object observed by the Machine Config Operator's build controller.
+	// +kubebuilder:validation:XValidation:rule="self >= oldSelf || (self == 0 && oldSelf > 0)", message="observedGeneration must not move backwards except to zero"
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Required
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// currentImagePullspec is the fully qualified image pull spec used by the MCO to pull down the new OSImage. This must include sha256.
```

Comment:
When is this field populated? After the build? We should mention when this value is updated/changes